src/iceberg/avro/avro_data_util_internal.h

```diff
@@ -26,10 +26,21 @@
 
 namespace iceberg::avro {
 
+/// \brief Append an Avro datum to an Arrow array builder.
+///
+/// This function handles schema evolution by using the provided projection to map
+/// fields from the Avro data to the expected Arrow schema.
+///
+/// \param avro_node The Avro schema node (must be a record at root level)
+/// \param avro_datum The Avro data to append
+/// \param projection Schema projection from `projected_schema` to `avro_node`
+/// \param projected_schema The projected schema
+/// \param array_builder The Arrow array builder to append to (must be a struct builder)
```

Comment:
So it's a undefined behavior to pass-in a different type builder? Can the argument here just a StructBuilder?

Reply:
That requires additional cast before using this function.